varnish-sys/src/vcl/ws_str_buffer.rs

```diff
@@ -46,184 +141,252 @@ impl<T> WsBuffer<'_, T> {
-                blob: data.as_ptr().cast(),
-                len: data.len(),
-                ..Default::default()
-            });
-            let next_available_byte = self.buffer.as_mut_ptr();
-            self.release(next_available_byte.offset_from(raw_start) as u32);
-
-            VCL_BLOB(info)
+            // force lifetime to be 'a because we are dropping self
+            let data = mem::transmute::<&[T], &'a [T]>(self.as_ref());
+            self.release(true);
+            RefCell::from(data)
         }
     }
 }
 
-impl<T> Write for WsBuffer<'_, T> {
-    fn write(&mut self, data: &[u8]) -> io::Result<usize> {
-        self.buffer.write(data)
-    }
-
-    fn flush(&mut self) -> io::Result<()> {
-        Ok(())
-    }
+#[expect(clippy::cast_ptr_alignment)]
+unsafe fn get_aligned_ptr_to_vrt_blob(buffer: *const u8) -> *const vrt_blob {
+    let offset = buffer.align_offset(align_of::<vrt_blob>());
+    buffer.add(offset).cast::<vrt_blob>()
 }
 
-impl<T> Drop for WsBuffer<'_, T> {
-    /// Ignore all the write commands, reclaiming the workspace memory
-    fn drop(&mut self) {
-        unsafe { self.release(0) };
-    }
+fn get_raw_start(ws: &ffi::ws) -> *const u8 {
+    ws.f.cast::<u8>()
 }
 
 #[cfg(test)]
 mod tests {
     use super::*;
-    use crate::vcl::TestWS;
-    use std::ffi::CStr;
+    use crate::vcl::{TestWS, Workspace};
+    use std::ffi::{CStr, CString};
 
     fn get_cstr(s: &VCL_STRING) -> &CStr {
         unsafe { CStr::from_ptr(s.0) }
     }
 
+    fn round_up_to_8(number: usize) -> usize {
+        number.div_ceil(8) * 8
+    }
+
+    fn get_ws_align<T>(ws: &mut Workspace) -> usize {
+        unsafe { (*ws.raw).f.align_offset(align_of::<T>()) }
+    }
+
+    fn buf_to_vec(buf: WsBlobBuffer) -> &[u8] {
+        let data = buf.finish();
+        let vrt_blob { blob, len, .. } = unsafe { *(data.0) };
+        unsafe { std::slice::from_raw_parts(blob.cast::<u8>(), len) }
+    }
+
     #[test]
     fn str_buffer() {
-        // init a workspace
         let mut test_ws = TestWS::new(160);
         let mut ws = test_ws.workspace();
 
         // first buffer call gets all available space
-        let mut buf = ws.get_str_buffer();
-        unsafe {
-            assert_eq!(160, (*ws.raw).r.offset_from((*ws.raw).f));
-        }
-        assert_eq!(buf.remaining(), 160);
-
-        // Validating writing
+        let mut buf = ws.get_str_buffer().unwrap();
+        assert_eq!(buf.remaining(), 159);
         buf.write_all(b"0123456789").unwrap();
-        let res = buf.finish();
-        assert_eq!(get_cstr(&res), c"0123456789");
+        assert_eq!(buf.remaining(), 149);
+        // saving 10 bytes + nul
+        assert_eq!(get_cstr(&buf.finish()), c"0123456789");
 
-        // second reservation get 160 - 10 bytes
-        let mut buf = ws.get_str_buffer();
-        assert_eq!(buf.remaining(), 150);
+        let mut buf = ws.get_str_buffer().unwrap();
+        assert_eq!(buf.remaining(), 160 - round_up_to_8(10 + 1) - 1);
```

Comment:
I think you are rounding up because of the alignment, which depends on the CPU arch?